modules/core/src/core/animation-loop.js

```diff
@@ -179,10 +176,10 @@ export default class AnimationLoop {
       this.gl.commit();
     }
 
-    if (this._canvasDataURLPromise) {
-      this._resolveCanvasDataURL(this.gl.canvas.toDataURL());
-      this._canvasDataURLPromise = null;
-      this._resolveCanvasDataURL = null;
+    if (this._resolveNextFrame) {
```

Comment:
I sense there is a subtle difference in that the old code happened inside the requestAnimationFrame timer (at least when `redraw` was not called "manually"), where as the new code is guaranteed to execute outside the timer callback.

Are there guarantees for when the default drawing buffer can be read when `preserveDrawingBuffers` is false that we should respect?

Reply:
The canvas contents are valid after draw operations complete and until the execution context (i.e. the entire call stack) in which the draw operations occurred returns.

As an aside, calling `redaw` outside an rAF callback would generally be a bad idea. Do you think it's being used that way?

Reply:
> The canvas contents are valid after draw operations complete and until the execution context (i.e. the entire call stack) in which the draw operations occurred returns.

As I understand it, to avoid subtle issues, `Promises` per spec do not fire until after control returns to the event loop, so this may not be rigorous.

> As an aside, calling redaw outside an rAF callback would generally be a bad idea. Do you think it's being used that way?

Yes, the "manual" `redraw()` method was recently added by @Pessimistress to support that use case, as a way to work around sync issues with independent webgl contexts in mapbox and google maps integrations.